compiler/qsc_circuit/src/circuit.rs

```diff
@@ -0,0 +1,673 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use rustc_hash::FxHashMap;
+use serde::Serialize;
+use std::{fmt::Display, fmt::Write, ops::Not, vec};
+
+/// Representation of a quantum circuit.
+/// Implementation of <https://github.com/microsoft/quantum-viz.js/wiki/API-schema-reference>
+#[derive(Clone, Serialize, Default, Debug, PartialEq)]
+pub struct Circuit {
+    pub operations: Vec<Operation>,
+    pub qubits: Vec<Qubit>,
+}
+
+#[derive(Clone, Serialize, Debug, PartialEq)]
+pub struct Operation {
+    #[allow(clippy::struct_field_names)]
+    pub gate: String,
+    #[serde(rename = "displayArgs")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub display_args: Option<String>,
+    #[serde(rename = "isControlled")]
+    #[serde(skip_serializing_if = "Not::not")]
+    pub is_controlled: bool,
+    #[serde(rename = "isAdjoint")]
+    #[serde(skip_serializing_if = "Not::not")]
+    pub is_adjoint: bool,
+    #[serde(rename = "isMeasurement")]
+    #[serde(skip_serializing_if = "Not::not")]
+    pub is_measurement: bool,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub controls: Vec<Register>,
+    pub targets: Vec<Register>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    pub children: Vec<Operation>,
+}
+
+#[derive(Serialize, Debug, Eq, Hash, PartialEq, Clone)]
+pub struct Register {
+    #[serde(rename = "qId")]
+    pub q_id: usize,
+    pub r#type: usize, // 0: quantum, 1: classical
+    #[serde(rename = "cId")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub c_id: Option<usize>,
+}
+
+#[derive(PartialEq, Clone, Serialize, Debug)]
+pub struct Qubit {
+    pub id: usize,
+    #[serde(rename = "numChildren")]
+    pub num_children: usize,
+}
+
+#[derive(Clone, Debug, Copy, Default)]
+pub struct Config {
+    pub no_qubit_reuse: bool,
```

Comment:
I'm really struggling with the name of this parameter. What it really means is "match base profile code generation". See `remapper.rs` for what that entails -- another struct whose naming I really struggled with.